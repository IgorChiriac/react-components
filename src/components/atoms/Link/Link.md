### Summary

Use `<Link />` to create hyperlinks to other web pages, files, locations within the same page, email addresses, or any other URL.

⚠️ &nbsp; Links are never headings

⚠️ &nbsp; Choose the color according to the following table:

| Background color     | Font Color                            |
| -------------------- | ------------------------------------- |
| white, grey or light | `colors.base.secondary` ( _default_ ) |
| dark                 | `colors.neutral.white`                |

### Examples

- Normal link

```js
import { Text, Link } from '@zopauk/react-components';

<Text size="large" as="p">
  Some text with
  <Link target="_blank" href="http://duckduckgo.com" onClick={() => alert('Link clicked!')}>
    a link
  </Link>
</Text>;
```

- Negative link

```js { "props": { "style": { "backgroundColor": "#00B9A7", "border": "none" } } }
import { Text, Link, colors } from '@zopauk/react-components';

<Text size="large" as="p" color={colors.neutral.white}>
  Some text with
  <Link
    color={colors.neutral.white}
    target="_blank"
    href="http://duckduckgo.com"
    onClick={() => alert('Link clicked!')}
  >
    a link
  </Link>
</Text>;
```

- With `target="_blank"`

```js
import { Text, Link } from '@zopauk/react-components';

<Text size="large" as="p">
  Some text with
  <Link target="_blank" href="http://duckduckgo.com" onClick={() => alert('Link clicked!')}>
    a link
  </Link>
</Text>;
```

- Without `target="_blank"` ( _notice that the square icon disappeared_ )

```js
import { Text, Link } from '@zopauk/react-components';

<Text size="large" as="p">
  Some text with
  <Link href="http://duckduckgo.com" onClick={() => alert('Link clicked!')}>
    a link
  </Link>
</Text>;
```
