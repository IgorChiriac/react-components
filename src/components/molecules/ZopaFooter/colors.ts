export const neutral900 = '#1C2139';
export const neutral200 = '#A9ACBA';
export const neutral400 = '#6F748B';
export const neutral800 = '#282F52';
