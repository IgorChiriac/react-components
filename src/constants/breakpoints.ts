export const breakpoints = {
  desktop: 1280,
  phone: 600,
  tablet: 720,
} as const;

export type TDeviceSizes = keyof typeof breakpoints;
