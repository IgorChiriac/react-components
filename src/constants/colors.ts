export type TTextHexColors = '#FFFFFF' | '#C5C6CD' | '#0D0A38' | '#1EC06A' | '#EE0505';
export type THeadingHexColors = '#FFFFFF' | '#0D0A38';
export type TLinkHexColors = '#FFFFFF' | '#0D0A38';

export const colors = {
  base: {
    primary: '#00B9A7',
    secondary: '#4A3EDE',
  },
  neutral: {
    white: '#FFFFFF',
    light: '#EAEAEE',
    medium: '#C5C6CD',
    dark: '#0D0A38',
  },
  semantic: {
    success: '#1EC06A',
    alert: '#FFB428',
    error: '#EE0505',
  },
};
