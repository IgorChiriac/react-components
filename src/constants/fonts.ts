export const openSans: string =
  '"Open Sans", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"';
export const alverata: string = 'Alverata, "Arial Black", Verdana, sans-serif';
