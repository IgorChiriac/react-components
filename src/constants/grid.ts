const grid = {
  breakpoints: {
    l: 992,
    m: 768,
    s: 576,
    xl: 1200,
    xs: 0,
  },
  cols: 12,
  gutter: 16,
  width: {
    l: 960,
    m: 720,
    s: 540,
    xl: 1140,
  },
};

export default grid;
