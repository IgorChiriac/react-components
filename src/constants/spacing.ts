export const spacing = [0, 4, 8, 16, 24, 32, 40, 56, 64, 104];
