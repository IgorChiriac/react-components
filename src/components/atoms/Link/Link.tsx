import React, { FC } from 'react';
import styled from 'styled-components';
import Text from '../Text/Text';
import { colors } from '../../../constants/colors';

export interface ILinkProps extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
  color?: string;
}

export interface ITargetIconProps extends React.SVGProps<SVGSVGElement> {}

const SLink = styled(Text)`
  cursor: pointer;
  text-decoration: none;
  user-select: none;
  appearance: none;

  color: ${({ color }) => color || colors.base.secondary};

  &:hover {
    opacity: 0.88;
  }

  &:active {
    opacity: 0.72;
  }
`;

const STargetIcon = styled.svg`
  margin-left: 6px;
  top: 2px;
  position: relative;
`;

const TargetIcon = (props: ITargetIconProps) => {
  return (
    <STargetIcon xmlns="http://www.w3.org/2000/svg" width="15" height="15">
      <g fill={props.color || colors.base.secondary}>
        <path d="M2.6 4H0v11h11v-2.6H2.6z" />
        <path d="M4 0v11h11V0H4zm9.7 1.3v8.4L5.2 1.3h8.5z" />
      </g>
    </STargetIcon>
  );
};

const Link: FC<ILinkProps> = React.forwardRef<HTMLAnchorElement, ILinkProps>((props, ref) => {
  const { children, ...rest } = props;
  return (
    <SLink ref={ref} weight="bold" forwardedAs="a" {...rest}>
      {children}
      {rest.target === '_blank' && <TargetIcon color={rest.color} />}
    </SLink>
  );
});

export default Link;
