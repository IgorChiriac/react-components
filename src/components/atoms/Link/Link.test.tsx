import { axe } from 'jest-axe';
import React from 'react';
import { render } from '@testing-library/react';
import { colors } from '../../../constants/colors';
import Link from './Link';

describe('<Link />', () => {
  it('renders the link with default values and the svg is hidden with no a11y violations', async () => {
    const { container } = render(<Link href="http://duckduckgo.com">text</Link>);
    expect(container.firstChild).toMatchSnapshot();

    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
  });

  it('renders the link with target _blank with the svg visible', () => {
    const { container } = render(
      <Link href="http://duckduckgo.com" target="_blank">
        text
      </Link>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it.each`
    hex                      | name
    ${colors.base.secondary} | ${'Blue'}
    ${colors.neutral.white}  | ${'White'}
  `('can render in different colors: $name â€“ $hex', ({ hex }) => {
    const { container } = render(
      <Link href="http://duckduckgo.com" color={hex}>
        text
      </Link>,
    );
    expect(container.firstChild).toHaveStyleRule('color', hex);
  });
});
