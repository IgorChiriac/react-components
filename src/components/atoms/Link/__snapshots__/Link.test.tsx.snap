// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> renders the link with default values and the svg is hidden with no a11y violations 1`] = `
.c0 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #4B3CFA;
}

.c0:hover {
  opacity: 0.88;
}

.c0:active {
  opacity: 0.72;
}

<a
  class="c0"
  href="http://duckduckgo.com"
>
  text
</a>
`;

exports[`<Link /> renders the link with target _blank with the svg visible 1`] = `
.c0 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #4B3CFA;
}

.c0:hover {
  opacity: 0.88;
}

.c0:active {
  opacity: 0.72;
}

.c1 {
  margin-left: 6px;
  top: 2px;
  position: relative;
}

<a
  class="c0"
  href="http://duckduckgo.com"
  target="_blank"
>
  text
  <svg
    class="c1"
    height="15"
    width="15"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="#4B3CFA"
    >
      <path
        d="M2.6 4H0v11h11v-2.6H2.6z"
      />
      <path
        d="M4 0v11h11V0H4zm9.7 1.3v8.4L5.2 1.3h8.5z"
      />
    </g>
  </svg>
</a>
`;
